### PR TITLE
Improve DateTime error handling and add some bad date tests

### LINKED ISSUE #112190: DateTimeException should be 4xx
An invalid MonthOfYear value (0) caused all shard to fail, RemoteTransportExeception and a `500` http status code. We feel this should rather be a code in the 4xx range, since it is an error in the search query.

```
Failed to execute phase [query], all shards failed; shardFailures {[PLWFic-nQKaByz8rQ1Labw][vendor-hotel][0]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-78c47d7586-8bzj7][100.64.110.2:9300][indices:data/read/search[phase/query]]
Caused by: java.time.DateTimeException: Invalid value for MonthOfYear (valid values 1 - 12): 0
	at java.base/java.time.temporal.ValueRange.checkValidValue(ValueRange.java:319)
	at java.base/java.time.temporal.ChronoField.checkValidValue(ChronoField.java:718)
	at java.base/java.time.LocalDate.of(LocalDate.java:273)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.time.DateTime.query(DateTime.java:50)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.time.DateFormatters.from(DateFormatters.java:2385)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.time.DateFormatters.from(DateFormatters.java:2368)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.time.JavaDateMathParser.parseDateTime(JavaDateMathParser.java:210)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.time.JavaDateMathParser.parse(JavaDateMathParser.java:63)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.mapper.DateFieldMapper$DateFieldType.parseToLong(DateFieldMapper.java:677)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.mapper.DateFieldMapper$DateFieldType.isFieldWithinQuery(DateFieldMapper.java:748)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.mapper.DateFieldMapper$DateFieldType.isFieldWithinQuery(DateFieldMapper.java:723)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.RangeQueryBuilder.getRelation(RangeQueryBuilder.java:474)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.RangeQueryBuilder.doSearchRewrite(RangeQueryBuilder.java:493)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.doRewrite(AbstractQueryBuilder.java:309)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.rewrite(AbstractQueryBuilder.java:280)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.BoolQueryBuilder.rewriteClauses(BoolQueryBuilder.java:400)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.BoolQueryBuilder.doRewrite(BoolQueryBuilder.java:356)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.rewrite(AbstractQueryBuilder.java:280)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.BoolQueryBuilder.rewriteClauses(BoolQueryBuilder.java:400)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.BoolQueryBuilder.doRewrite(BoolQueryBuilder.java:355)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.AbstractQueryBuilder.rewrite(AbstractQueryBuilder.java:280)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SubSearchSourceBuilder.rewrite(SubSearchSourceBuilder.java:89)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SubSearchSourceBuilder.rewrite(SubSearchSourceBuilder.java:37)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.Rewriteable.rewrite(Rewriteable.java:57)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.Rewriteable.rewrite(Rewriteable.java:40)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.Rewriteable.rewrite(Rewriteable.java:125)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SearchSourceBuilder.rewrite(SearchSourceBuilder.java:1176)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.builder.SearchSourceBuilder.rewrite(SearchSourceBuilder.java:93)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.Rewriteable.rewrite(Rewriteable.java:57)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.Rewriteable.rewrite(Rewriteable.java:40)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.internal.ShardSearchRequest$RequestRewritable.rewrite(ShardSearchRequest.java:587)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.internal.ShardSearchRequest$RequestRewritable.rewrite(ShardSearchRequest.java:576)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.index.query.Rewriteable.rewrite(Rewriteable.java:57)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.SearchService.createSearchContext(SearchService.java:1147)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.SearchService.createContext(SearchService.java:1076)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:685)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:559)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:78)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:75)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:100)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:991)
	at org.elasticsearch.server@8.16.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
```

### PR DESCRIPTION
**Pull request by smalyshev**
Catch any `DateTimeException` as some may not be caught during parsing phase and only come up during `from` phase. 

Also add some bad date tests. 

Fixes https://github.com/elastic/elasticsearch/issues/112190

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/time/JavaDateMathParser.java [hunk 1]
```diff
@@ -219,7 +219,7 @@ private Instant parseDateTime(String value, ZoneId timeZone, boolean roundUpIfNo
 
                 return DateFormatters.from(accessor).withZoneSameLocal(timeZone).toInstant();
             }
-        } catch (IllegalArgumentException | DateTimeParseException e) {
+        } catch (IllegalArgumentException | DateTimeException e) {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/common/time/JavaDateMathParser.java:222]: In the [main parse method](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java#L185) we just catch `Exception`, and throw `IllegalArgumentException` - could we just do the same here?

File: server/src/test/java/org/elasticsearch/index/mapper/DateFieldTypeTests.java [hunk 2]
```diff
@@ -142,6 +143,46 @@ private void doTestIsFieldWithinQuery(DateFieldType ft, DirectoryReader reader,
...
+            ElasticsearchParseException.class,
+            () -> ft.isFieldWithinQuery(reader, "2015-00-01", "2016-04-03", randomBoolean(), randomBoolean(), zone, null, context)
+        );
+        assertThrows(
+            ElasticsearchParseException.class,
+            () -> ft.isFieldWithinQuery(reader, "2015-01-01", "2016-04-00", randomBoolean(), randomBoolean(), zone, null, context)
+        );
+        assertThrows(
+            ElasticsearchParseException.class,
+            () -> ft.isFieldWithinQuery(reader, "2015-22-01", "2016-04-00", randomBoolean(), randomBoolean(), zone, null, context)
```
Comment (craigtaverner) [inline server/src/test/java/org/elasticsearch/index/mapper/DateFieldTypeTests.java:157]: I see tests for `0` month and `0` days. I'm curious what happens with `0` years?

Comment (smalyshev) [reply server/src/test/java/org/elasticsearch/index/mapper/DateFieldTypeTests.java:157]: I think 0 years should be valid? I mean there's no actual year 0 in UTC/Gregorian, but Java treats year 0 as 1 BCE and it's a valid date I think.

**Comment by javanna** [comment]
@smalyshev looks like this is a valuable change, are there concerns with it? Do you still plan to get it in?

**Comment by smalyshev** [comment]
@elasticmachine update branch

**Comment by smalyshev** [comment]
@javanna I think it's ok just needs to be reviewed/approved.

File: server/src/main/java/org/elasticsearch/common/time/JavaDateMathParser.java (hunk 1 above)
Comment (smalyshev) [reply server/src/main/java/org/elasticsearch/common/time/JavaDateMathParser.java:223]: I'm not sure, catching `Exception` kinda makes me feel uneasy usually, is it ok in this context? Usually I prefer to specify.

**Review by javanna — APPROVED** [review]

**Pull request merged**